Scylla node benchmarks

KnifeyMoloko opened the pull request:
This PR adds the ability to perform node benchmarks on our Ubuntu db nodes. The 2 benchmarks added are `sysbench` for cpu tests and `cassandra-fio` for storage benchmarking. More info can be found in the readme doc added by this PR.


Trello tasl: https://trello.com/c/BmhZ5DN2
Example longevity job: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/mc-longevity/job/mc-byo-new-test/115
Elasticsearch index for collected benchmark stats: `node_benchmarks`

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+            runner.run_benchmarks()
+
+    @staticmethod
+    def _collect_benchmark_output():
```
It looks like as "abstract" method which must be redefined. So, I guess, it should have "@abc.abstractmethod" decorator.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+        return core_num
+
+    @staticmethod
+    def run_fio():
```
The same here -> @abc.abstractmethod

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+        """
+        This class is meant to install and run benchmarking
+        tools on given cluster nodes and collect the output.
+        """
```
This docstring must be defined for class, not one of it's methods.

KnifeyMoloko replied:
Fixed

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+                self._node.install_package(pkg)
+            return True
+        except Exception:
+            LOGGER.warning("Failed to install Ubuntu prerequisites for the node benchmarking tools.")
```
Why is it "warning" log message knowing that the next step is to "raise" an exception?

KnifeyMoloko replied:
Fixed (removed the exception reraising)

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+            LOGGER.debug("Build and install sysbench errors: %s", result.stderr)
+            return True
+        except Exception:
+            LOGGER.warning("Failed to install sysbench from source.")
```
Same here.

KnifeyMoloko replied:
Fixed

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+        return False
+
+    def run_sysbench(self, thread_num: int = None, test_time: int = 120) -> str:
+        thread_num = thread_num if thread_num else self._get_number_of_cores_on_node()
```
It may be shorter:

```
thread_num = thread_num or self._get_number_of_cores_on_node()
```

KnifeyMoloko replied:
Fixed

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+            LOGGER.warning("Failed to clone sysbench repo.")
+            raise
+
+    def _check_sysbench_installation(self) -> bool:
```
Name doesn't correspond the logic. Logic returns "bool" but name doesn't imply it.
So, need either to rename it or stop returning boolean values.

For xample:
```def _is_sysbench_installed(self): -> bool:```

KnifeyMoloko replied:
Fixed

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+    def _clone_sysbench_repo(self):
+        try:
+            LOGGER.info("Cloning sysbench repo...")
+            clone_cmd = "git clone https://github.com/akopytov/sysbench.git"
```
Define this var before try block and use it in all the log messages.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+        if sysbench_run_result.stderr:
+            LOGGER.warning("sysbench benchmark ran with errors: \n%s", sysbench_run_result.stderr)
+        LOGGER.info("Sysbench results:")
+        LOGGER.info("%s", sysbench_run_result.stdout)
```
```
LOGGER.info("Sysbench results:\n%s", sysbench_run_result.stdout)
```

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,137 @@
+        cat_cmd = "cat /proc/cpuinfo"
+        regex_pattern = re.compile(r"cpu cores\s+:\s+(?P<core_num>\d+)")
+        cat_result = self._remoter.run(cat_cmd).stdout
+        core_num = int(regex_pattern.search(cat_result).groupdict()["core_num"])
```
It doesn't look safe. If for some reason search don't find anything then we get python errors such as 'KeyError' and 'TypeError'.

KnifeyMoloko replied:
I agree that it's not safe, but I'd rather have it fail in a case where it can't get the data.

aleksbykov replied:
We have property on basenode 
   @property
    def cpu_cores
Why not use it?

aleksbykov replied:
Why not use cpu_cores property?

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -12,13 +12,55 @@
+    latency_min: float = field(init=False)
+    latency_max: float = field(init=False)
+    latency_avg: float = field(init=False)
+    latency_95: float = field(init=False)
```
I guess better to add 'p' symbol as it is done in lots of other places:
```
95th percentile -> p95
99th percentile -> p99
```
As a result:
```
latency_p95: float = field(init=False)
```

KnifeyMoloko replied:
Agreed, better to keep the "p". Fixed.

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -12,13 +12,55 @@
+            r"execution time \(avg/stddev\):\s+(?P<time_avg>[\d.]+)/(?P<time_stdev>[\d.]+)"
+        )
+        LOGGER.info("Running regex on:")
+        LOGGER.info(self.cmd_output)
```
```
 LOGGER.info("Running regex on:\n%s", self.cmd_output)
```

KnifeyMoloko replied:
Fixed

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -12,13 +12,55 @@
+        # self.thread_fairness_avg = eps_regex.search(self.cmd_output).groupdict()["thread_avg"]
+        # self.thread_fairness_stdev = eps_regex.search(self.cmd_output).groupdict()["thread_stdev"]
+        # self.thread_fairness_time_avg = eps_regex.search(self.cmd_output).groupdict()["time_avg"]
+        # self.thread_fairness_time_stdev = eps_regex.search(self.cmd_output).groupdict()["time_stdev"]
```
Any helpful comment for this block of 'commented out' code?

KnifeyMoloko replied:
Fixed with rebase (the commented out code was not present in the latter PR commits, now it won't show up in the earlier ones either)

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -42,6 +84,7 @@ def install_benchmark_tools(self):
             runner.install_benchmark_tools()
 
     def run_benchmarks(self):
+        # ParallelObject(self._benchmark_runners).run(lambda x: x.run_benchmarks())
```
?

KnifeyMoloko replied:
Fixed with rebase (the commented out code was not present in the latter PR commits, now it won't show up in the earlier ones either)

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -60,13 +103,14 @@ def __init__(self, node: "BaseNode"):
 
     def install_benchmark_tools(self) -> bool:
-        self._clone_sysbench_repo()
+        self._clone_repo("https://github.com/akopytov/sysbench.git")
```
This repo path is written more than 1 time in different places. It is not correct.
Please, reuse one single place for it. Moreover, the repo you use is personal for some guy. 
It is better to use our own clone which gets synced periodically.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -129,9 +173,27 @@ def _get_number_of_cores_on_node(self) -> int:
+        if cassandra_fio_run_result.stderr:
+            LOGGER.warning("cassandra-fio benchmark ran with errors: \n%s", cassandra_fio_run_result.stderr)
+        LOGGER.info("cassandra-fio results:")
+        LOGGER.info("%s", cassandra_fio_run_result.stdout)
```
Same here. Could be one line

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -129,9 +173,27 @@ def _get_number_of_cores_on_node(self) -> int:
+
+
+if __name__ == "__main__":
+    so = Path("sysbench_out")
```
Please, look at our doc located at [1].
According to it "so" is not good naming. But, moreover, it is used just once, so it could be used just in place.

[1] https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI/edit#heading=h.6byizauykwn9

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -129,9 +173,27 @@ def _get_number_of_cores_on_node(self) -> int:
-    @staticmethod
-    def run_fio():
-        ...
+    def run_fio(self):
```
Seeing this change I think that it must have been part of the first commit in this chain. 
It makes no sense to update the code you make in the same PR like this.

KnifeyMoloko replied:
It was hard not to update it, since these are the initial commits for this code. That being said, I did rebase the commits a bit to make it less of hassle to review the changes.

vponomaryov commented on sdcm/cluster.py:
```diff
@@ -3363,6 +3363,15 @@ def get_all_tables_with_cdc(self, db_node: BaseNode) -> List[str]:
+        self.node_benchmark_manager.install_benchmark_tools()
+        self.node_benchmark_manager.run_benchmarks()
+        self.node_benchmark_manager.print_benchmark_results_to_log()
+
```
Please, place original method in the same place to reduce number of changed code lines.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/sct_config.py:
```diff
@@ -1220,8 +1220,7 @@ class SCTConfiguration(dict):
              type=boolean,
-             help="Flag for running db node benchmarks before the tests")
-
+             help="Flag for running db node benchmarks before the tests"),
```
Addition of this missing comma must not be here. It should be added in the previous commit.
Need to reduce code line changes as much as possible when you do multi-commit PR.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -150,13 +152,21 @@ def _get_db_node_instance_type(self):
 
     def _install_ubuntu_prerequisites(self) -> bool:
         package_list = ["make", "automake", "libtool", "pkg-config", "libaio-dev", "fio"]
+        time_1 = time.time()
```
Please, name it more self-explanatory [1].

[1] https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI/edit#heading=h.6byizauykwn9

KnifeyMoloko replied:
Fixed with rebase

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -150,13 +152,21 @@ def _get_db_node_instance_type(self):
-                self._node.install_package(pkg)
+                self._node.install_package(pkg, wait_for_package_manager=False)
+
+            time_2 = time.time()
```
Same here.

KnifeyMoloko replied:
Fixed with rebase

vponomaryov reviewed (changes requested):
Please see inline comments.
Also, general comment:

Please, make commits be aligned with each other. Do not make it fix each other. 
Such thing makes it be obfuscation on following steps:
- code review
- git commit history inspection anytime later after PR merge.

KnifeyMoloko commented:
> Please see inline comments. Also, general comment:
> 
> Please, make commits be aligned with each other. Do not make it fix each other. Such thing makes it be obfuscation on following steps:
> 
> * code review
> * git commit history inspection anytime later after PR merge.

@vponomaryov Thanks for the comment. I went ahead and rebased the commit history. I think the changes make it a bit more sensible now.

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+
+    def _build_and_install_sysbench(self) -> bool:
+        # pylint: disable=broad-except
+        build_and_install_script = """\
+            cd ./sysbench
+            ./autogen.sh
+            ./configure --without-mysql
+            make -j
+            make install
```
Do we really want to build it on each separate test run? Maybe we want to reuse it from S3? Picking up latest version we put there?
Even if not add it now, but at least, add a "TODO" comment and create new task in our task tracker.
What do you think?

KnifeyMoloko replied:
It sounds good in theory. My fear is that: a) downloading from S3 might take longer than the current process b) we're adding maintenance cost by this. That being said, I created a Trello taks for this: https://trello.com/c/KIbuX25o and we'll see how it goes.

fruch replied:
@aleksbykov docker isn't installed on DB nodes (installing docker would take more time)

@KnifeyMoloko why not taking the official binary of it ? (instead of building it)

https://packagecloud.io/akopytov/sysbench/install#bash-deb

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+        """
+        try:
+            self._remoter.sudo(shell_script_cmd(build_and_install_script))
+            return True
```
Do you really need boolean here? It's result is not checked directly. Moreover, method name doesn't imply it.
For now, it looks like golang approach, not pythonic.

KnifeyMoloko replied:
Fixed (removed)

vponomaryov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+
+    def run_benchmarks(self):
+        parallel = ParallelObject(self._benchmark_runners, timeout=1200)
+        parallel.run(lambda x: x.run_benchmarks())
```
This block of code gets changed in both other commits of this PR. Can you please combine these changes in one commit?

KnifeyMoloko replied:
Fixed (moved the change to the first commit)

KnifeyMoloko commented on sdcm/sct_config.py:
```diff
@@ -1218,7 +1218,9 @@ class SCTConfiguration(dict):
              type=int,
              help="Number of iops for ebs type io2|io3|gp3"),
-
+        dict(name="run_node_benchmarks", env="SCT_RUN_NODE_BENCHMARKS",
```
Good idea, it will be more clear. Fixed.

KnifeyMoloko commented on sdcm/tester.py:
```diff
@@ -3054,3 +3054,7 @@ def is_shard_awareness_driver(self) -> bool:
+
+    def run_node_benchmarks(self):
+        cluster: BaseScyllaCluster = self.db_cluster
+        cluster.node_benchmark_manager.install_benchmark_tools()
```
Nope, fixed (removed)

KnifeyMoloko commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+    def print_benchmark_results_to_log(self):
+        results = self._collect_benchmark_output()
+        LOGGER.info("Benchmark results:")
+        LOGGER.info(results)
```
I felt gracious with log lines there. Fixed (inlined to 1 line).

KnifeyMoloko commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+
+    def _get_db_node_instance_type(self):
+        backend = self._node.parent_cluster.params.get("cluster_backend")
+        if backend == "aws":
```
Fixed

KnifeyMoloko commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,252 @@
+            self._remoter.sudo(clone_cmd)
+            LOGGER.debug("Finished cloning from %s.", repo_url)
+        except ParallelObjectException as exc:
+            LOGGER.warning("Failed to clone from %s. Failed with: %s", repo_url, exc)
```
Good idea. Fixed (moved and modified to `sdcm/utils/git.py`)

KnifeyMoloko commented on sdcm/cluster_aws.py:
```diff
@@ -778,7 +778,6 @@ def __init__(self, ec2_ami_id, ec2_subnet_id, ec2_security_group_ids,  # pylint:
         cluster_uuid = self.test_config.test_id()
         cluster_prefix = cluster.prepend_user_prefix(user_prefix, 'db-cluster')
         node_prefix = cluster.prepend_user_prefix(user_prefix, 'db-node')
-
```
Fixed

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+# Node Benchmarks for Scylla
+
+Whether you're running performance jobs or investigating an issue, you might
+want to now if the hardware Scylla is running meets you baseline perfromance
+expectations. That's what node benchmarks are for.
```
`if the hardware Scylla is running meets you baseline perfromance expectations.` ->
if the hardware Scylla is running **ON** meets you**R** baseline per**FOR**mance expectations.

KnifeyMoloko replied:
Fixed

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+want to now if the hardware Scylla is running meets you baseline perfromance
+expectations. That's what node benchmarks are for.
+
+### What benchmarks are we running?
```
I think it must be written in `present simple` tense:
`Which benchmarks do we run?`
The reason for it is that we do not mean some specific benchmark which runs `now`, but those which we are able to run any time.

KnifeyMoloko replied:
Good catch - it does sound better in present simple. Fixed.

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+
+### What benchmarks are we running?
+
+We're running:
```
Same here

KnifeyMoloko replied:
Fixed

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+
+### How to run the benchmarks?
+
+Simply add the line `run_node_benchmarks: true` in your test case yaml.
```
This one refers to the old option name. Need to update it.

KnifeyMoloko replied:
Fixed

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+
+### At what point in the test are the benchmarks performed?
+
+Both benchmarks are performed at the end of adding nodes to the test cluster
+(nodes added after the cluster initialization are not included).
```
Saying "the test cluster" you refer to some specific object. So, in current case it must be "a test Scylla cluster":

`Both benchmarks get performed at the end of "adding nodes" step to a test Scylla cluster, note that nodes which get added after the "cluster init" step won't be benchmarked.`

KnifeyMoloko replied:
Fixed (with a slightly different wording)

vponomaryov commented on docs/node_benchmarks.md:
```diff
@@ -0,0 +1,351 @@
+
+Simply add the line `run_node_benchmarks: true` in your test case yaml.
+
+### At what point in the test are the benchmarks performed?
```
`At what point in the test are the benchmarks performed?` doesn't sound correct to me
What about this one:
`At which point in time of a test run these benchmarks get performed?`
?

KnifeyMoloko replied:
Fixed with: `At which point in time of a test run are these benchmarks performed?`

roydahan commented on sdcm/utils/git.py:
```diff
@@ -23,3 +23,16 @@ def get_git_current_branch() -> str:
     return proc.stdout.decode(encoding="utf-8").strip()
+
+
+def clone_repo(remoter, repo_url: str):
```
Would be nice by adding such a util, to also refactor another place that does the same and can earn from this.
(In a separate commit).
This is one of the ways to make sure other people are using the common utils.

KnifeyMoloko replied:
I refactored 2 places in which `git clone` was used directly (`jepsen_test.py`, `cluster.py`) in the final commit to this PR

roydahan commented:
Can you please point me to a sample elasticsearch results you have for your runs?

roydahan commented on sdcm/cluster.py:
```diff
@@ -3361,6 +3369,14 @@ def get_all_tables_with_cdc(self, db_node: BaseNode) -> List[str]:
+
+        self.node_benchmark_manager.add_nodes(self.nodes)
+        self.node_benchmark_manager.install_benchmark_tools()
+        self.node_benchmark_manager.run_benchmarks()
```
I think that one of the main things here is missing.
The comparison between the nodes and sending a warning event (for now).
I don't know what are the most relevant metrics we would like to follow, but we need one or more metrics that we care about and compare.

KnifeyMoloko replied:
Agreed that a comparison and warning can be useful. That being said I'd rather do the node fairness comparison and a run fairness comparison (node metrics for the same type of instance across runs) in a separate PR. It would be easier to review on it's own.

roydahan replied:
We can postpone this one to the next PR, just open a task for it.

KnifeyMoloko commented:
> Can you please point me to a sample elasticsearch results you have for your runs?

@roydahan the ES index used is called `node_benchmarks`.  There are also example of how the results are structured in the `node_benchmarks.md` doc file in this PR.

roydahan commented:
> > Can you please point me to a sample elasticsearch results you have for your runs?
> 
> @roydahan the ES index used is called `node_benchmarks`. There are also example of how the results are structured in the `node_benchmarks.md` doc file in this PR.

As we've seen in the ealsticsearch, the data model doesn't fit the way we want to handle the results, we need to flatten the JSON to 1-2 levels tops.
(Also, remove the exisiting docs and recreate the index pattern after we insert the new ones).

aleksbykov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,240 @@
+    def run_sysbench(self, thread_num: int = None, test_time: int = 120):
+        thread_num = thread_num or self._get_number_of_cores_on_node()
+        run_cmd = f"sysbench cpu --threads={thread_num} --time={test_time} run"
+        sysbench_run_result = self._remoter.run(run_cmd)
```
does it require root privileges to run?
Also it is good to add ignore_status, so if the command faile, it will not crash test

KnifeyMoloko replied:
It does not require root privileges. +1 to the `ignore_status` idea, I'll add that once I'm done with the ES doc restructuring.

KnifeyMoloko replied:
I decided to wrap the execution and getting the results in a try / except clause, so the `ignore_status` should no be relevant anymore.

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -1221,7 +1221,9 @@ class SCTConfiguration(dict):
              type=int,
              help="Number of iops for ebs type io2|io3|gp3"),
-
+        dict(name="run_db_node_benchmarks", env="SCT_RUN_DB_NODE_BENCHMARKS",
+             type=boolean,
```
Did you add description to config.md for new parameter?

KnifeyMoloko replied:
Good catch :) I'll add the description for that.

KnifeyMoloko replied:
Description for the param added.

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -3021,6 +3027,7 @@ def __init__(self, cluster_uuid=None, cluster_prefix='cluster', node_prefix='nod
             self.add_nodes(n_nodes, enable_auto_bootstrap=self.auto_bootstrap)
         else:
             raise ValueError('Unsupported type: {}'.format(type(n_nodes)))
+        self.run_node_benchmarks()
```
in doc file, you mentioned that this will be run only during cluster initialization, but call run benchmark in __init__, So the benchmark will be executed for each node which is added during nemesis, won't be?

KnifeyMoloko replied:
It shouldn't - it's the `__init__` method of the `Base Cluster` class, so my understanding is it will only be called when a cluster is being created. If a nemesis class creates an additional cluster, then yes, it would call the benchmark methods on that additional cluster.

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -3498,6 +3513,7 @@ def __init__(self, *args, **kwargs):
         self.nemesis_threads = []
         self.nemesis_count = 0
         self.test_config = TestConfig()
+        self.node_benchmark_manager = ScyllaClusterBenchmarkManager()
```
why do you need same on cluster level? may be rename to self.cluster_benchmark_manager?

aleksbykov replied:
or you can init it here (even it as singleton) and in basenode get reference from self.parent_cluster.self.node_benchmark_manager

KnifeyMoloko replied:
Good catch. This looks to be redundant - I removed it and left the `ScyllaClusterBenchmarkManager` only in the `BaseCluster` class.

aleksbykov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,233 @@
+            make install
+        """
+        try:
+            self._remoter.sudo(shell_script_cmd(build_and_install_script))
```
maybe use ignore_status?

KnifeyMoloko replied:
Yup, fixed by using `ignore_status`.

aleksbykov commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,233 @@
+        fio_reports_path = "/home/ubuntu/cassandra-fio/reports/"
+        cat_cmd = f"cat {fio_reports_path}{'lcs.64k.fio.json'}"
+        try:
+            cat_out = self._remoter.run(cat_cmd)
```
you can use ignore_status and don't use so wide try except. What do you think?

KnifeyMoloko replied:
Good idea in most places in this module (I fixed it in 2 places), but possibly not here, as it's not just the `cat` command that could fail, but also the following JSON manipulation. I think it's safer to wrap it in a `try / except` to avoid it crashing a test without good reason.

aleksbykov commented:
@KnifeyMoloko resolve comments please

KnifeyMoloko commented:
> @KnifeyMoloko resolve comments please

@aleksbykov Thanks for the comments - I fixed all of them. (The logic for getting the number of cpu's is now replaced by the call to the nodes `.cpu_cores` property too.)

aleksbykov reviewed (approved):
LGTM

KnifeyMoloko commented:
@roydahan Hi Roy, I think we can merge this PR (all comments were addressed and I've resolved all conflicts).

roydahan commented:
@fruch can you please review as well?

fruch commented on sdcm/utils/benchmarks.py:
```diff
@@ -0,0 +1,220 @@
+
+    def install_benchmark_tools(self):
+        clone_repo(self._remoter, "https://github.com/akopytov/sysbench.git")
+        clone_repo(self._remoter, "https://github.com/KnifeyMoloko/cassandra-fio.git")
```
please add a comment here, why we are using a private fork. (and a link to the PR to upstream, if any)

KnifeyMoloko replied:
@fruch I moved to using my own fork here sine the upstream repo was not updated since 2017 and it had a few small issues I wanted to fix for own needs (e.g. making report format default to JSON). I've added a comment with the upstream repo's URL.

fruch commented:
@KnifeyMoloko @roydahan aren't we adding this to any job, so it would start running regularly (and we can start having real results) ?

fruch reviewed (dismissed):
LGTM

KnifeyMoloko commented:
> @KnifeyMoloko @roydahan aren't we adding this to any job, so it would start running regularly (and we can start having real results) ?

That's a good question - @roydahan I can add the benchmarks to test case yamls as part of this task: https://trello.com/c/KLMKltuX We just need to decide what runs we want to add benchmarks to.

roydahan commented:
> 

We can decide on which jobs to enable it in the next task.
